File: src/main/Java/boj_step21/boj4779.java
import java.io.*;

public class boj4779 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        try {
            while (true) {
                String s = "-";
                int n = Integer.parseInt(reader.readLine());
                if (n == 0) writer.append(s + "\n");
                else writer.append(hyphens(s, n) + "\n").flush();
            }
        } catch (Exception e) { e.getMessage(); }
    }
    private static String hyphens(String s,  int n) {
        if (n == 0) return s;
        String temp = "";
        for (int i = 0; i < 3; i++) {
            if (i == 1) temp += s.replace("-", " ");
            else temp += s;
        }
        return hyphens(temp, n - 1);
    }
}

File: src/main/boj_step21/boj4779.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    try {
        while (true) {
            val s = "-"
            val n = r.readLine().toInt()
            if (n == 0) w.appendLine(s)
            else {
                w.appendLine(hyphens(s, n))
                w.flush()
            }
        }
    } catch (e: Exception) { e.message }
}

private fun hyphens(s:String, n: Int): String {
    if (n == 0) return s
    var temp = ""
    repeat(3) {
        when (it) {
            1 -> s.forEach { temp += " " }
            else -> temp += s
        }
    }
    return hyphens(temp, n - 1)
}